Method or class for composing more than 1 free wires into one - Using DiagramRewriter class

ACE07-Sev opened the pull request:
Greetings,

Hope you are well. My apologies, I had to make a new PR because the other one couldn't be accessed locally anymore. My apologies once more.

I have made the amendments requested by @dimkart. The rewrite_diagram.py is now merged into rewrite.py. I have kept a copy of the old one just in case you would like to compare it in the future, I frankly think a separate file would have been better as it would have been its own rewrite package, and would not have stood out by being in the rewrite.py which operates at box level. Regardless, please let me know if there's anything else I need to fix.

Addresses #106 . Continues #110 .

Thommy257 commented on tests/test_rewrite.py:
```diff
@@ -172,3 +173,26 @@ def test_curry_functor():
+
+    rewritten_diagram = MergeWiresRewriter().rewrite_diagram(diagram)
+
+    assert rewritten_diagram == expected_diagram
```
end with new line

Thommy257 commented on lambeq/rewrite/base.py:
```diff
@@ -126,6 +133,34 @@ def __call__(self, box: Box) -> Diagram | None:
         return self.rewrite(box) if self.matches(box) else None
 
 
+class DiagramRewriter(ABC):
```
put in new file

Thommy257 commented on lambeq/rewrite/base.py:
```diff
@@ -318,6 +353,16 @@ def rewrite(self, box: Box) -> Diagram:
         return new_box
 
 
+class MergeWiresRewriter(DiagramRewriter):
```
put in new file

ACE07-Sev commented:
Thommy dear I had done that initially, but Dr. Kartsaklis requested I put all inside rewrite's base.py.

dimkart commented:
> Thommy dear I had done that initially, but Dr. Kartsaklis requested I put all inside rewrite's base.py.

I haven't suggested such a thing, check my comment above. I asked for a new file diagram_rewrite.py.

ACE07-Sev commented:
My sincerest apologies then, I misunderstood what you mentioned. However, wouldn't it better to have a separate folder with a base.py and init.py to match rewrite's current structure? 

If not, I'll proceed to add the rewrite_diagram.py to the rewrite folder, but that would also mean I have to merge the init and base of rewrite_diagram folder I initially made into a single file correct?

ACE07-Sev commented:
Dr @dimkart , I have made the file as requested. I hope I didn't make any other bad misunderstandings. My apologies in that regard.

ACE07-Sev commented:
@Thommy257 I have made two branches, in the two PRs we have at the moment. One does it using a new folder with its own init and base, and one is as Dr. Kartsaklis requested which is the new file rewrite_diagram.py (if I didn't misunderstand again, if so my apologies in advance).

I have kept both for comparison. Please let me know which one you prefer as well. Thank you so much.

dimkart commented on lambeq/rewrite/__init__.py:
```diff
@@ -13,7 +13,10 @@
 from lambeq.rewrite.base import (RewriteRule, CoordinationRewriteRule,
                                  CurryRewriteRule, SimpleRewriteRule, Rewriter)
+
+from lambeq.rewrite.rewrite_diagram import (DiagramRewriter, MergeWiresRewriter)
```
New line.

ACE07-Sev replied:
By new line you mean to add another empty line after 21?

nikhilkhatri replied:
Yes, all files should end with a newline.

dimkart commented on lambeq/rewrite/rewrite_diagram.py:
```diff
@@ -0,0 +1,80 @@
+        return not diagram.cod == S
+
+    def rewrite_diagram(self, diagram: Diagram) -> Diagram:
+        return (diagram >> Box('MERGE', diagram.cod, S))
```
New line.

dimkart commented on lambeq/rewrite/rewrite_diagram.py:
```diff
@@ -0,0 +1,80 @@
+    """A rewrite rule for imperative sentences."""
+
+    def matches(self, diagram: Diagram) -> bool:
+        return not diagram.cod == S
```
That's not a good condition, please check the length of the codomain type (it has to be > 1).

dimkart commented on lambeq/__init__.py:
```diff
@@ -113,7 +115,8 @@
                               is_pregroup_diagram, remove_cups, remove_swaps)
 from lambeq.rewrite import (RewriteRule, CoordinationRewriteRule,
-                            CurryRewriteRule, SimpleRewriteRule, Rewriter)
+                            CurryRewriteRule, SimpleRewriteRule, Rewriter,
```
Let's sort the imports alphabetically.

ACE07-Sev replied:
So you mean like : 
CoordinationRewriteRule, CurryRewriteRule, DiagramRewriter, MergeWiresRewriter, Rewriter, RewriteRule, SimpleRewriteRule ?

nikhilkhatri replied:
Imports are typically grouped together if they're subclasses of the same base class (multiple `RewriteRule`s for eg.), and then ordered alphabetically.

I'd recommend taking a look at some of the init files in lambeq, for eg. https://github.com/CQCL/lambeq/blob/main/lambeq/__init__.py

dimkart commented on lambeq/rewrite/__init__.py:
```diff
@@ -13,7 +13,10 @@
 
 __all__ = ['RewriteRule', 'CoordinationRewriteRule', 'CurryRewriteRule',
-           'SimpleRewriteRule', 'Rewriter']
+           'SimpleRewriteRule', 'Rewriter', 'DiagramRewriter',
```
Let's sort.

dimkart commented on lambeq/rewrite/rewrite_diagram.py:
```diff
@@ -0,0 +1,80 @@
+    def matches(self, diagram: Diagram) -> bool:
+        return not diagram.cod == S
+
+    def rewrite_diagram(self, diagram: Diagram) -> Diagram:
```
Please pass an argument here to define the output type, with default `S`.

dimkart commented on lambeq/rewrite/rewrite_diagram.py:
```diff
@@ -0,0 +1,80 @@
+        """Check if the given diagram should be rewritten."""
+
+    @abstractmethod
+    def rewrite_diagram(self, diagram: Diagram) -> Diagram:
```
We would also need a method that applies to a list of diagrams, e.g. `rewrite_diagrams`.

ACE07-Sev replied:
I was wondering, what about a single method that given the diagram type being Diagram or list of Diagram, it would automatically perform the rewriting correctly and returning either a diagram or a list of diagrams.

ACE07-Sev replied:
My understanding is that we rely on the __call__ method correct? No one should run rewrite_diagram because then it wouldn't check for matches.

nikhilkhatri replied:
Regarding your second comment, from looking at the code It isnt clear to me whether `DiagramRewriter` is meant to be analogous to `RewriteRule` or `Rewriter`. 
In the current rewriting interface, the user does not call the `RewriteRule` on a box directly. This is the responsibility of the `Rewriter`, which is initialised with a list of `RewriteRule`s, and then when called on a diagram, applies these rules to it.
It might make sense to have a similar structure for the whole-diagram rewrites

ACE07-Sev replied:
Exactly, it's supposed to be a merged version of RewriteRule and Rewriter since we do it at diagram level. I was initially going to do it with a init and base format, but was requested to instead to it in a single file, so I followed the current structure. 

I actually agree that it should be a bit more similar to the current rewrite rule and rewriter instance, but maybe that's not possible given we are performing the rewriting at diagram level? Functorially I mean.

dimkart replied:
I don't think it makes much sense to have a single rewriter with many rules at the diagram level, similar to what we do at the box level. Box rules are local, but diagram "rules" are global to the diagram, so it doesn't make sense to apply more than one of them. And mixing box rules with diagram rules is not a good idea since the scope is so different.

dimkart commented:
@ACE07-Sev Good work -- I've left a few comments. 

I have to repeat though, please stop opening new PRs for every change you want to apply!! We need all comments for a specific piece of work to be in one place. Please from now on make all your changes in this PR.

ACE07-Sev commented:
I didn't mean to Dr. Kartsaklis. As I explained to Thommy in the open PR, because I renamed the branch, it got deleted, so I opened the same branch's PR, and it created a duplicate.

My apologies regardless. I'll amend the items you kindly mentioned.

dimkart commented on lambeq/rewrite/rewrite_diagram.py:
```diff
@@ -0,0 +1,83 @@
+# limitations under the License.
+
+"""
+Rewrite
```
Let's update this comment to something like:

```
Diagram Rewrite
============
Class hierarchy for allowing rewriting at the diagram level  (as opposed 
to rewrite rules that apply on the box level).

Subclass :py:class:'DiagramRewriter' to define a custom rewrite rule.
```

ACE07-Sev commented:
dear @dimkart , have you had time to review the new addition sir? In addition to your kind comments, I added a rather small addition to the __call__ methods for the base.py classes to add support for multiple diagrams. It now also is more similar to the multiple diagram support I added for the rewrite_diagram.py class. 

Please let me know what you think!

dimkart commented:
Hi, if you can rebase your code to the new main branch it would be great. Some minor comments might still follow, but this looks good in general. Thanks for your work!

ACE07-Sev commented:
Done sir.

dimkart commented:
@ACE07-Sev Hi, thanks -- nothing else to be done for now, at some point someone from the team will will push a few minor adjustments.
